### ISSUE TITLE
Consider the Samples Report PDF on the backend

### ISSUE THREAD

**Issue by ysbaddaden** [instedd/cdx#1843]
The samples report PDF in #1840 relies on JavaScript libraries on the frontend, which is different from how we generate PDF elsewhere on the system. Ideally we shouldn't introduce many ways to do the same thing.

The only bothering thing is that the charts are generated on the frontend using d3 (which we already used to generate other charts). Ideally we'd generate them on the backend, but I can't find any good Ruby library to generate SVG. Running d3 on the backend feels weird, and it doesn't work with MiniRacer (I get some missing JS references for d3).

Follow up of #1840

**Comment by ysbaddaden** [comment]
I tried running `d3` on the backend, using `domjs` to simulate a browser. It works with Node16 , but it doesn't work with Node10 or MiniRacer (v8 embedded in Ruby) because of missing `TextEncoder` implementation which prevents loading `domjs` to run.

Too bad, it would have been fun to be able to run d3 on the server.

**Comment by ysbaddaden** [comment]
I actually got `d3` + `jsdom` running on node10 (default on debian buster), with the following `package.json`:

```json
{
  "dependencies": {
    "d3": "^3.5.16",
    "jsdom": "^16.7.0",
    "text-encoding-polyfill": "^0.6.7",
    "util": "^0.12.5"
  }
}
```

Then the following script is working for example:

```js
const JSDOM = require("jsdom").JSDOM
const d3 = require("d3")

const html = "<!DOCTYPE html><html><body></body></html>"
const document = new JSDOM(html).window.document

const svg = d3.select(document.body).append("svg")
console.log(svg.node().outerHTML)
```

for node16 we can skip the `util` and `text-encoding-polyfill` packages, and directly:

```js
import { JSDOM } from "jsdom"
import d3 from "d3"

const html = "<!DOCTYPE html><html><body></body></html>"
const document = new JSDOM(html).window.document

const svg = d3.select(document.body).append("svg")
console.log(svg.node().outerHTML)
```

**Comment by ysbaddaden** [comment]
I also got it running inside `mini_racer` but it's a little bit too hacky for my taste, and requires to bundle `jsdom` (`browserify -r jsdom`):

```ruby
require "mini_racer"
ctx = MiniRacer::Context.new

# jsdom + polyfills
ctx.eval <<JAVASCRIPT
var window = global = this;
function setTimeout(fn, timeout) {}
function clearTimeout(tid) {}
JAVASCRIPT
ctx.eval File.read("node_modules/text-encoding-polyfill/lib/encoding-indexes.js")
ctx.eval File.read("node_modules/text-encoding-polyfill/lib/encoding.js")
ctx.eval File.read("vendor/assets/javascripts/jsdom_browserify.min.js")

# d3
ctx.eval File.read("node_modules/d3/d3.min.js")

# actual script
script = <<~JAVASCRIPT
const jsdom = require("jsdom");
const dom = new jsdom.JSDOM("<!DOCTYPE html><html><body></body></html>");

const svg = d3
  .select(dom.window.document.body)
  .append("svg")

svg.node().outerHTML
JAVASCRIPT

svg = ctx.eval(script)
p svg
```

**Comment by ysbaddaden** [comment]
We convened with @leandroradusky to replace the React components with a few JavaScript functions to generate the SVG charts using `d3` only. These functions would then be called on the frontend directly (still passing data and configuration using [gon](https://github.com/gazay/gon)) yet allow the backend to also call these functions to generate the SVG (see above) to be rendered in Prawn through [prawn-svg](https://github.com/mogest/prawn-svg).

**Comment by ysbaddaden** [comment]
We implemented an in-between solution for the time being: we send the SVG charts rendered in the browser to the backend that will inject them into the PDF file.

To further improve on the solution we could extract each chart as individual javascript files, that can be loaded in the frontend _and_ run through nodejs calls on the backend (using d3 and jsdom as explained above), so the backend would be autonomous to render SVG charts in whatever contexts, especially those _without_ a browser (e.g. sending emails, background jobs, ...).

**Comment by leandroradusky** [comment]
Just dropping a question here: will those charts-as-javascript-files be reactive to, i.e. threshold updates for the challenge purpose plots in a straight manner? Here the react state becomes handy and I guess this can be done but I'm not sure how sophisticated the solution needs to be.

**Comment by ysbaddaden** [comment]
Yes. It's just a function that you call instead of inlining the d3 calls right into the React component.

**Comment by sardar-usman** [comment]
@leandroradusky It is working fine.
